File: shipley_checks/src/Stage_4/guild_scorer_rust/src/metrics/m5_beneficial_fungi.rs
//! METRIC 5: BENEFICIAL FUNGI NETWORKS (MYCORRHIZAE & ENDOPHYTES)
//!
//! **PHASE 3 OPTIMIZATION**: Pre-filtered LazyFrame with column projection
//!
//! Scores Common Mycorrhizal Networks and individual fungal associations
//! using shared organism counting and coverage analysis.
//!
//! **Memory optimization**:
//!   - Old: Receives full fungi_df (11,711 rows), filters inside count_shared_organisms
//!   - New: Receives pre-filtered fungi_lazy (7 rows), selects only 5 needed columns
//!   - Reuses same fungi_lazy as M3 and M4 (triple reuse!)
//!
//! **Columns needed** (M5 selects these 5):
//!   1. plant_wfo_id - Plant identification (for filtering)
//!   2. amf_fungi - Arbuscular mycorrhizal fungi
//!   3. emf_fungi - Ectomycorrhizal fungi
//!   4. endophytic_fungi - Endophytic fungi
//!   5. saprotrophic_fungi - Saprotrophic fungi
//!
//! R reference: shipley_checks/src/Stage_4/metrics/m5_beneficial_fungi.R

use polars::prelude::*;
use anyhow::Result;
use rustc_hash::FxHashMap;
use crate::utils::{Calibration, percentile_normalize, count_shared_organisms, materialize_with_columns, filter_to_guild};

/// Column requirements for M5 calculation (from fungi parquet)
pub const REQUIRED_FUNGI_COLS: &[&str] = &[
    "plant_wfo_id",
    "amf_fungi",
    "emf_fungi",
    "endophytic_fungi",
    "saprotrophic_fungi",
];

/// Result of M5 calculation
#[derive(Debug)]
pub struct M5Result {
    /// Combined network + coverage score (0-1 scale)
    pub raw: f64,
    /// Percentile score (0-100, HIGH = GOOD)
    pub norm: f64,
    /// Network connectivity score
    pub network_score: f64,
    /// Fraction of plants with beneficial fungi
    pub coverage_ratio: f64,
    /// Number of shared fungi
    pub n_shared_fungi: usize,
    /// Number of plants with beneficial fungi
    pub plants_with_fungi: usize,
    /// Map of fungus_name → plant_count for detailed analysis
    pub fungi_counts: FxHashMap<String, usize>,
}

/// Calculate M5: Beneficial Fungi Networks
///
/// **PHASE 3 OPTIMIZATION**: Reuses fungi LazyFrame from scorer, filters after column projection
///
/// R reference: m5_beneficial_fungi.R::calculate_m5_beneficial_fungi
pub fn calculate_m5(
    plant_ids: &[String],        // Guild plant IDs for filtering
    fungi_lazy: &LazyFrame,      // Schema-only scan (from scorer, reused from M3/M4!)
    calibration: &Calibration,
) -> Result<M5Result> {
    // STEP 1: Materialize fungi columns and filter to guild
    let fungi_df = materialize_with_columns(
        fungi_lazy,
        REQUIRED_FUNGI_COLS,
        "M5 fungi",
    )?;

    let guild_fungi = filter_to_guild(&fungi_df, plant_ids, "plant_wfo_id", "M5")?;

    let n_plants = guild_fungi.height();

    // Beneficial fungi columns for counting
    let columns = &["amf_fungi", "emf_fungi", "endophytic_fungi", "saprotrophic_fungi"];

    // Get plant IDs from the filtered DataFrame (now local to this function)
    let guild_plant_ids: Vec<String> = guild_fungi
        .column("plant_wfo_id")?
        .str()?
        .into_iter()
        .filter_map(|opt| opt.map(|s| s.to_string()))
        .collect();

    // Count shared beneficial fungi (fungi hosted by ≥2 plants)
    // guild_fungi is already filtered (7 rows), so this just analyzes the guild
    let beneficial_counts = count_shared_organisms(&guild_fungi, &guild_plant_ids, columns)?;

    // COMPONENT 1: Network score (weight 0.6)
    // For each shared fungus, calculate network connectivity
    let mut network_raw = 0.0;
    for (_org_name, count) in &beneficial_counts {
        if *count >= 2 {
            network_raw += *count as f64 / n_plants as f64;
        }
    }

    // COMPONENT 2: Coverage ratio (weight 0.4)
    // What fraction of plants have ANY beneficial fungi?
    let plants_with_beneficial = count_plants_with_beneficial_fungi(&guild_fungi, &guild_plant_ids, columns)?;
    let coverage_ratio = plants_with_beneficial as f64 / n_plants as f64;

    // Combined score: 60% network, 40% coverage
    let p5_raw = network_raw * 0.6 + coverage_ratio * 0.4;

    // Percentile normalize
    let m5_norm = percentile_normalize(p5_raw, "p3", calibration, false)?;

    Ok(M5Result {
        raw: p5_raw,
        norm: m5_norm,
        network_score: network_raw,
        coverage_ratio,
        n_shared_fungi: beneficial_counts.len(),
        plants_with_fungi: plants_with_beneficial,
        fungi_counts: beneficial_counts,
    })
}

/// Count how many plants have at least one beneficial fungus
fn count_plants_with_beneficial_fungi(
    fungi_df: &DataFrame,
    plant_ids: &[String],
    columns: &[&str],
) -> Result<usize> {
    use rustc_hash::FxHashSet;

    let plant_id_set: FxHashSet<&String> = plant_ids.iter().collect();
    let plant_col = fungi_df.column("plant_wfo_id")?.str()?;

    let mut count = 0;
    for idx in 0..fungi_df.height() {
        if let Some(plant_id) = plant_col.get(idx) {
            if !plant_id_set.contains(&plant_id.to_string()) {
                continue; // Not in guild
            }

            // Check if this plant has any beneficial fungi
            let mut has_beneficial = false;
            for col_name in columns {
                if let Ok(col) = fungi_df.column(col_name) {
                    // Phase 0-4 parquets use Arrow list columns
                    if let Ok(list_col) = col.list() {
                        if let Some(list_series) = list_col.get_as_series(idx) {
                            if list_series.len() > 0 {
                                has_beneficial = true;
                                break;
                            }
                        }
                    } else if let Ok(str_col) = col.str() {
                        // Fallback: pipe-separated strings (legacy format)
                        if let Some(fungi_str) = str_col.get(idx) {
                            if !fungi_str.is_empty() {
                                has_beneficial = true;
                                break;
                            }
                        }
                    }
                }
            }

            if has_beneficial {
                count += 1;
            }
        }
    }

    Ok(count)
}

#[cfg(test)]
mod tests {
    use approx::assert_relative_eq;

    #[test]
    fn test_network_score_calculation() {
        // If 2 plants out of 3 share a fungus, connectivity = 2/3 ≈ 0.667
        // If 3 plants out of 3 share a fungus, connectivity = 3/3 = 1.0

        let n_plants = 3.0_f64;
        let mut network_raw = 0.0_f64;

        // Fungus 1 shared by 2 plants
        network_raw += 2.0 / n_plants;
        // Fungus 2 shared by 3 plants
        network_raw += 3.0 / n_plants;

        assert_relative_eq!(network_raw, 1.667, epsilon = 0.01);
    }

    #[test]
    fn test_coverage_ratio() {
        // If 2 out of 3 plants have any beneficial fungi
        let plants_with_fungi = 2;
        let n_plants = 3;
        let coverage_ratio = plants_with_fungi as f64 / n_plants as f64;

        assert_relative_eq!(coverage_ratio, 0.667, epsilon = 0.01);
    }

    #[test]
    fn test_combined_score() {
        // network_raw = 1.0, coverage_ratio = 1.0
        let p5_raw = 1.0_f64 * 0.6 + 1.0 * 0.4;
        assert_relative_eq!(p5_raw, 1.0, epsilon = 0.0001);

        // network_raw = 0.5, coverage_ratio = 0.8
        let p5_raw = 0.5_f64 * 0.6 + 0.8 * 0.4;
        assert_relative_eq!(p5_raw, 0.62, epsilon = 0.0001);
    }
}

File: shipley_checks/src/Stage_4/metrics/m5_beneficial_fungi.R
#!/usr/bin/env Rscript
#
# METRIC 5: BENEFICIAL FUNGI NETWORKS (MYCORRHIZAE & ENDOPHYTES)
#
# ============================================================================
# ECOLOGICAL RATIONALE
# ============================================================================
#
# Beneficial fungi provide critical ecosystem services:
#
# 1. ARBUSCULAR MYCORRHIZAL FUNGI (AMF):
#    - Form symbiotic associations with ~80% of plant species
#    - Extend root systems via hyphal networks
#    - Enhance nutrient uptake (especially phosphorus)
#    - Improve drought resistance
#    - Create underground networks connecting multiple plants (Common Mycorrhizal Networks)
#
# 2. ECTOMYCORRHIZAL FUNGI (EMF):
#    - Form associations with trees (especially conifers, oaks, beeches)
#    - Sheath root tips in fungal mantle
#    - Critical for nutrient cycling in forest ecosystems
#
# 3. ENDOPHYTIC FUNGI:
#    - Live inside plant tissues without causing disease
#    - Produce defensive compounds (alkaloids, terpenoids)
#    - Enhance stress tolerance
#    - Some provide systemic acquired resistance
#
# 4. SAPROTROPHIC FUNGI:
#    - Decompose organic matter
#    - Recycle nutrients from dead plant material
#    - Improve soil structure
#
# KEY CONCEPT: Common Mycorrhizal Networks (CMNs)
# When multiple plants share the same beneficial fungi, they form underground
# networks that:
# - Transfer resources between plants (carbon, nitrogen, water)
# - Communicate defense signals
# - Support establishment of new seedlings
# - Stabilize plant communities
#
# ============================================================================
# CALCULATION STEPS
# ============================================================================
#
# STEP 1: Count shared beneficial fungi across guild members
#   - For each fungal taxon, count how many plants host it
#   - Only fungi shared by ≥2 plants contribute to network score
#   - Rationale: Single-plant fungi don't create networks
#
# STEP 2: Network score (weight: 0.6)
#   - For each shared fungus, calculate: count / n_plants
#   - This gives the "network connectivity" (0-1 scale)
#   - Sum across all shared fungi
#   - Interpretation:
#     * High score: Many fungi shared by most plants (strong CMNs)
#     * Low score: Few fungi, or fungi shared by only 2-3 plants
#
# STEP 3: Coverage ratio (weight: 0.4)
#   - Count how many plants have ANY beneficial fungi
#   - coverage_ratio = plants_with_fungi / total_plants
#   - Rationale: Even without shared networks, individual fungal associations
#     provide plant benefits (nutrient uptake, stress tolerance)
#
# STEP 4: Combined score
#   - p5_raw = network_score × 0.6 + coverage_ratio × 0.4
#   - This balances network effects (60%) with individual benefits (40%)
#
# STEP 5: Percentile normalization
#   - Uses Köppen tier-stratified calibration for metric 'p3'
#   - HIGH p5_raw → HIGH percentile (good fungal networks)
#   - Note: No final inversion needed (high percentile = high score)
#
# ============================================================================
# DATA SOURCES
# ============================================================================
#
# fungi_df (fungal_guilds_pure_r.csv):
# - Columns: plant_wfo_id, amf_fungi, emf_fungi, endophytic_fungi, saprotrophic_fungi
# - Each column contains pipe-separated GBIF fungal taxon IDs
# - Derived from FunGuild, GlobalFungi, and literature mining
#
# ============================================================================
# PARITY REQUIREMENTS
# ============================================================================
#
# To maintain 100% parity with Python scorer:
# 1. Must use same count_shared_organisms logic (≥2 plants threshold)
# 2. Must use same network score formula (count / n_plants)
# 3. Must use same coverage ratio calculation
# 4. Must use same weights (0.6 network, 0.4 coverage)
# 5. Must use same Köppen tier calibration file for 'p3' metric
#
# Python implementation: guild_scorer_v3.py lines 1420-1477
#
# ============================================================================


#' Calculate M5: Beneficial Fungi Networks
#'
#' Scores Common Mycorrhizal Networks and individual fungal associations
#' using shared organism counting and coverage analysis.
#'
#' @param plant_ids Character vector of WFO taxon IDs for the guild
#' @param guild_plants Data frame with plant metadata (not used in M5)
#' @param fungi_df Data frame with plant-fungi associations
#' @param count_shared_organisms_fn Function to count shared organisms
#' @param percentile_normalize_fn Function to convert raw score to percentile
#'
#' @return List with three elements:
#'   - raw: Combined network + coverage score (0-1 scale)
#'   - norm: Percentile score (0-100, HIGH = GOOD)
#'   - details: List with network_score, coverage_ratio, n_shared_fungi,
#'              plants_with_fungi
#'
calculate_m5_beneficial_fungi <- function(plant_ids,
                                          guild_plants,
                                          fungi_df,
                                          count_shared_organisms_fn,
                                          percentile_normalize_fn) {

  n_plants <- length(plant_ids)

  # Count shared beneficial fungi (fungi hosted by ≥2 plants)
  beneficial_counts <- count_shared_organisms_fn(
    fungi_df, plant_ids,
    'amf_fungi', 'emf_fungi', 'endophytic_fungi', 'saprotrophic_fungi'
  )

  # COMPONENT 1: Network score (weight 0.6)
  # For each shared fungus, calculate network connectivity
  network_raw <- 0
  for (org_name in names(beneficial_counts)) {
    count <- beneficial_counts[[org_name]]
    if (count >= 2) {
      network_raw <- network_raw + (count / n_plants)
    }
  }

  # COMPONENT 2: Coverage ratio (weight 0.4)
  # What fraction of plants have ANY beneficial fungi?
  guild_fungi <- fungi_df %>% dplyr::filter(plant_wfo_id %in% plant_ids)
  plants_with_beneficial <- 0

  if (nrow(guild_fungi) > 0) {
    for (i in seq_len(nrow(guild_fungi))) {
      row <- guild_fungi[i, ]
      has_beneficial <- FALSE
      for (col in c('amf_fungi', 'emf_fungi', 'endophytic_fungi', 'saprotrophic_fungi')) {
        col_val <- row[[col]]
        if (is.list(col_val)) col_val <- col_val[[1]]
        if (!is.null(col_val) && length(col_val) > 0 && !all(is.na(col_val))) {
          has_beneficial <- TRUE
          break
        }
      }
      if (has_beneficial) {
        plants_with_beneficial <- plants_with_beneficial + 1
      }
    }
  }

  coverage_ratio <- plants_with_beneficial / n_plants

  # Combined score: 60% network, 40% coverage
  p5_raw <- network_raw * 0.6 + coverage_ratio * 0.4

  # Percentile normalize
  m5_norm <- percentile_normalize_fn(p5_raw, 'p3')

  # -------------------------------------------------------------------------
  # STEP 6: Build fungi counts by category (for network profile generation)
  # -------------------------------------------------------------------------
  # Build fungi_counts: fungus_name → number of plants hosting it
  # Also track category for each fungus (AMF, EMF, Endophytic, Saprotrophic)

  fungi_counts <- list()
  fungus_category_map <- list()  # fungus_name → category string

  for (i in seq_len(nrow(guild_fungi))) {
    row <- guild_fungi[i, ]

    # AMF fungi
    amf <- row$amf_fungi[[1]]
    if (!is.null(amf) && length(amf) > 0) {
      for (fungus in amf) {
        if (is.null(fungi_counts[[fungus]])) {
          fungi_counts[[fungus]] <- 0
          fungus_category_map[[fungus]] <- "AMF"
        }
        fungi_counts[[fungus]] <- fungi_counts[[fungus]] + 1
      }
    }

    # EMF fungi
    emf <- row$emf_fungi[[1]]
    if (!is.null(emf) && length(emf) > 0) {
      for (fungus in emf) {
        if (is.null(fungi_counts[[fungus]])) {
          fungi_counts[[fungus]] <- 0
          fungus_category_map[[fungus]] <- "EMF"
        }
        fungi_counts[[fungus]] <- fungi_counts[[fungus]] + 1
      }
    }

    # Endophytic fungi
    endo <- row$endophytic_fungi[[1]]
    if (!is.null(endo) && length(endo) > 0) {
      for (fungus in endo) {
        if (is.null(fungi_counts[[fungus]])) {
          fungi_counts[[fungus]] <- 0
          fungus_category_map[[fungus]] <- "Endophytic"
        }
        fungi_counts[[fungus]] <- fungi_counts[[fungus]] + 1
      }
    }

    # Saprotrophic fungi
    sapro <- row$saprotrophic_fungi[[1]]
    if (!is.null(sapro) && length(sapro) > 0) {
      for (fungus in sapro) {
        if (is.null(fungi_counts[[fungus]])) {
          fungi_counts[[fungus]] <- 0
          fungus_category_map[[fungus]] <- "Saprotrophic"
        }
        fungi_counts[[fungus]] <- fungi_counts[[fungus]] + 1
      }
    }
  }

  list(
    raw = p5_raw,
    norm = m5_norm,
    fungi_counts = fungi_counts,
    fungus_category_map = fungus_category_map,
    details = list(
      network_score = network_raw,
      coverage_ratio = coverage_ratio,
      n_shared_fungi = length(beneficial_counts),
      plants_with_fungi = plants_with_beneficial
    )
  )
}
